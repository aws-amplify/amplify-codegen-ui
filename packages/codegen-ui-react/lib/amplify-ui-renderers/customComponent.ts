/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import { StudioComponentChild } from '@aws-amplify/codegen-ui';
import { JsxChild, JsxElement, factory } from 'typescript';
import { isAliased, removeAlias } from '../helpers';
import { ReactComponentRenderer } from '../react-component-renderer';

export default class CustomComponentRenderer<TPropIn> extends ReactComponentRenderer<TPropIn> {
  renderElement(renderChildren: (children: StudioComponentChild[]) => JsxChild[]): JsxElement {
    const children = this.component.children ? this.component.children : [];
    const element = factory.createJsxElement(
      this.renderOpeningElement(),
      renderChildren(children),
      factory.createJsxClosingElement(factory.createIdentifier(this.component.componentType)),
    );

    if (isAliased(this.component.componentType)) {
      this.importCollection.addImport(
        `./${removeAlias(this.component.componentType)}`,
        `${removeAlias(this.component.componentType)} as ${this.component.componentType}`,
      );
    } else {
      this.importCollection.addImport(`./${this.component.componentType}`, 'default');
    }

    return element;
  }
}
