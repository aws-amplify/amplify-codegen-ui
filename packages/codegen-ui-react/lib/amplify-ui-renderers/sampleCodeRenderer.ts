/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import { BaseComponentProps } from '@aws-amplify/ui-react';

import { factory, JsxAttribute, JsxAttributeLike, JsxSelfClosingElement } from 'typescript';
import { FrontendManagerRendererConstants } from '@aws-amplify/codegen-ui';
import { ReactComponentRenderer } from '../react-component-renderer';

export default class SampleCodeRenderer extends ReactComponentRenderer<BaseComponentProps> {
  renderElement(): JsxSelfClosingElement {
    const tagName = this.component.name ?? FrontendManagerRendererConstants.unknownName;
    // const prop = new Map<string, BoundFrontendManagerComponentProperty>();
    // this.collectExposedProps(this.component, prop);

    const propsArray: JsxAttribute[] = [];
    /*  TODO:  move over to boundProperties
    const defaultValue = 'defaultValue';
    const defaultValueExpr = factory.createJsxExpression(undefined, factory.createIdentifier(defaultValue));
    props?.forEach((value, key) => {
      if (value.exposedAs) {
        const displayExpr =
          value.value !== undefined ? factory.createStringLiteral(value.value.toString()) : defaultValueExpr;
        const attr = factory.createJsxAttribute(factory.createIdentifier(key), displayExpr);
        propsArray.push(attr);
      }
    });
    */

    return factory.createJsxSelfClosingElement(
      factory.createIdentifier(tagName),
      undefined,
      factory.createJsxAttributes(propsArray),
    );
  }

  /*
  private collectExposedProps(
    component: FrontendManagerComponent,
    collected: Map<string, FrontendManagerComponentProperty>,
  ) {
    const moreItems = Object.entries(component.properties).filter((m) => !(m[1].exposedAs == null));
    moreItems?.forEach((value, index) => {
      if (!collected.has(value[0])) {
        collected.set(value[0], value[1]);
      }
    });

    component.children?.forEach((child) => {
      this.collectExposedProps(child, collected);
    });
  }
  */

  private addExposedPropAttributes(attributes: JsxAttributeLike[], tagName: string) {
    const propsAttr = factory.createJsxSpreadAttribute(factory.createIdentifier('props'));
    attributes.push(propsAttr);

    const overrideAttr = factory.createJsxSpreadAttribute(
      factory.createCallExpression(factory.createIdentifier('getOverrideProps'), undefined, [
        factory.createPropertyAccessExpression(
          factory.createIdentifier('props'),
          factory.createIdentifier('overrides'),
        ),
        factory.createStringLiteral(tagName),
      ]),
    );
    attributes.push(overrideAttr);
  }

  mapProps(props: BaseComponentProps): BaseComponentProps {
    return props;
  }
}
