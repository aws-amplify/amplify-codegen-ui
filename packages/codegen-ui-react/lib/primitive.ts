/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import { FixedStudioComponentProperty } from '@aws-amplify/codegen-ui';
import { factory, SyntaxKind, TypeParameterDeclaration, TypeNode } from 'typescript';

export type PrimitiveLevelPropConfiguration<ConfigType> = {
  [componentType: string]: { [eventType: string]: ConfigType };
};

export enum Primitive {
  Alert = 'Alert',
  Badge = 'Badge',
  Button = 'Button',
  ButtonGroup = 'ButtonGroup',
  Card = 'Card',
  CheckboxField = 'CheckboxField',
  Collection = 'Collection',
  Divider = 'Divider',
  Expander = 'Expander',
  ExpanderItem = 'ExpanderItem',
  Flex = 'Flex',
  Grid = 'Grid',
  Heading = 'Heading',
  Icon = 'Icon',
  Image = 'Image',
  Link = 'Link',
  Loader = 'Loader',
  Menu = 'Menu',
  MenuButton = 'MenuButton',
  MenuItem = 'MenuItem',
  Pagination = 'Pagination',
  PasswordField = 'PasswordField',
  PhoneNumberField = 'PhoneNumberField',
  Placeholder = 'Placeholder',
  Radio = 'Radio',
  RadioGroupField = 'RadioGroupField',
  Rating = 'Rating',
  ScrollView = 'ScrollView',
  SearchField = 'SearchField',
  SelectField = 'SelectField',
  SliderField = 'SliderField',
  StepperField = 'StepperField',
  SwitchField = 'SwitchField',
  Table = 'Table',
  TableBody = 'TableBody',
  TableCell = 'TableCell',
  TableFoot = 'TableFoot',
  TableHead = 'TableHead',
  TableRow = 'TableRow',
  Tabs = 'Tabs',
  TabItem = 'TabItem',
  Text = 'Text',
  TextAreaField = 'TextAreaField',
  TextField = 'TextField',
  ToggleButton = 'ToggleButton',
  ToggleButtonGroup = 'ToggleButtonGroup',
  View = 'View',
  VisuallyHidden = 'VisuallyHidden',
  Autocomplete = 'Autocomplete',
}

/*
 * temporary list of Primitives with a change event. Final implementation will pull from amplify UI
 */
export const PrimitivesWithChangeEvent: Set<Primitive> = new Set([
  Primitive.CheckboxField,
  Primitive.PasswordField,
  Primitive.PhoneNumberField,
  Primitive.RadioGroupField,
  Primitive.SearchField,
  Primitive.SelectField,
  Primitive.SliderField,
  Primitive.StepperField,
  Primitive.SwitchField,
  Primitive.TextAreaField,
  Primitive.TextField,
]);

export const PrimitiveChildrenPropMapping: Partial<Record<Primitive, string>> = {
  [Primitive.Alert]: 'label',
  [Primitive.Badge]: 'label',
  [Primitive.Button]: 'label',
  [Primitive.Heading]: 'label',
  [Primitive.Link]: 'label',
  [Primitive.MenuButton]: 'label',
  [Primitive.MenuItem]: 'label',
  [Primitive.Radio]: 'label',
  [Primitive.TableCell]: 'label',
  [Primitive.Text]: 'label',
  [Primitive.ToggleButton]: 'label',
};

export const PrimitiveDefaultPropertyValue: PrimitiveLevelPropConfiguration<FixedStudioComponentProperty> = {
  [Primitive.CheckboxField]: {
    checked: { value: false, type: 'boolean' },
  },
  [Primitive.PasswordField]: {
    value: { value: '', type: 'string' },
  },
  [Primitive.PhoneNumberField]: {
    value: { value: '', type: 'string' },
  },
  [Primitive.RadioGroupField]: {
    value: { value: '', type: 'string' },
  },
  [Primitive.SearchField]: {
    value: { value: '', type: 'string' },
  },
  [Primitive.SwitchField]: {
    isChecked: { value: false, type: 'boolean' },
  },
  [Primitive.TextAreaField]: {
    value: { value: '', type: 'string' },
  },
  [Primitive.TextField]: {
    value: { value: '', type: 'string' },
  },
};

export function isPrimitive(componentType: string): boolean {
  return Object.values(Primitive).includes(componentType as Primitive);
}

export const PrimitiveTypeParameter: Partial<
  Record<Primitive, { declaration: () => TypeParameterDeclaration[] | undefined; reference: () => TypeNode[] }>
> = {
  [Primitive.TextField]: {
    declaration: () => [
      factory.createTypeParameterDeclaration(
        factory.createIdentifier('Multiline'),
        factory.createKeywordTypeNode(SyntaxKind.BooleanKeyword),
        undefined,
      ),
    ],
    reference: () => [factory.createTypeReferenceNode(factory.createIdentifier('Multiline'), undefined)],
  },
  [Primitive.Collection]: {
    declaration: () => undefined,
    reference: () => [factory.createKeywordTypeNode(SyntaxKind.AnyKeyword)],
  },
};

export const PRIMITIVE_OVERRIDE_PROPS = 'PrimitiveOverrideProps';
/**
 * export declare type PrimitiveOverrideProp<T> = Partial<T> & React.DOMAttributes<HTMLDivElement>;
 */
export const primitiveOverrideProp = factory.createTypeAliasDeclaration(
  undefined,
  [factory.createModifier(SyntaxKind.ExportKeyword), factory.createModifier(SyntaxKind.DeclareKeyword)],
  factory.createIdentifier(PRIMITIVE_OVERRIDE_PROPS),
  [factory.createTypeParameterDeclaration(factory.createIdentifier('T'), undefined, undefined)],
  factory.createIntersectionTypeNode([
    factory.createTypeReferenceNode(factory.createIdentifier('Partial'), [
      factory.createTypeReferenceNode(factory.createIdentifier('T'), undefined),
    ]),
    factory.createTypeReferenceNode(
      factory.createQualifiedName(factory.createIdentifier('React'), factory.createIdentifier('DOMAttributes')),
      [factory.createTypeReferenceNode(factory.createIdentifier('HTMLDivElement'), undefined)],
    ),
  ]),
);
