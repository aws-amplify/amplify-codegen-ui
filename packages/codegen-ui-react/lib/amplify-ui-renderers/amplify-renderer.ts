/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import { StudioNode, StudioComponent, StudioComponentChild } from '@aws-amplify/codegen-ui';
import { JsxElement, JsxFragment, JsxSelfClosingElement } from 'typescript';
// add primitives in alphabetical order
import {
  AlertProps,
  BadgeProps,
  ButtonProps,
  ButtonGroupProps,
  CardProps,
  CheckboxFieldProps,
  DividerProps,
  ExpanderProps,
  ExpanderItemProps,
  FlexProps,
  GridProps,
  HeadingProps,
  IconProps,
  ImageProps,
  LinkProps,
  LoaderProps,
  MenuItemProps,
  MenuProps,
  PaginationProps,
  PasswordFieldProps,
  PhoneNumberFieldProps,
  PlaceholderProps,
  RadioProps,
  RadioGroupFieldProps,
  RatingProps,
  ScrollViewProps,
  SearchFieldProps,
  SelectFieldProps,
  SliderFieldProps,
  StepperFieldProps,
  SwitchFieldProps,
  TabItemProps,
  TabsProps,
  TableCellProps,
  TableBodyProps,
  TableFootProps,
  TableHeadProps,
  TableProps,
  TableRowProps,
  TextAreaFieldProps,
  TextFieldProps,
  ToggleButtonProps,
  ToggleButtonGroupProps,
  ViewProps,
  VisuallyHiddenProps,
  TextProps,
} from '@aws-amplify/ui-react';
import { Primitive } from '../primitive';
import { ReactStudioTemplateRenderer } from '../react-studio-template-renderer';
import CustomComponentRenderer from './customComponent';
import CollectionRenderer from './collection';
import { ReactComponentRenderer } from '../react-component-renderer';

export class AmplifyRenderer extends ReactStudioTemplateRenderer {
  renderJsx(
    component: StudioComponent | StudioComponentChild,
    parent?: StudioNode,
  ): JsxElement | JsxFragment | JsxSelfClosingElement {
    const node = new StudioNode(component, parent);
    const renderChildren = (children: StudioComponentChild[]) => children.map((child) => this.renderJsx(child, node));

    // add Primitive in alphabetical order
    switch (component.componentType) {
      case Primitive.Alert:
        return new ReactComponentRenderer<AlertProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.Badge:
        return new ReactComponentRenderer<BadgeProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.Button:
        return new ReactComponentRenderer<ButtonProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.ButtonGroup:
        return new ReactComponentRenderer<ButtonGroupProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.Card:
        return new ReactComponentRenderer<CardProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.CheckboxField:
        return new ReactComponentRenderer<CheckboxFieldProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.Collection:
        return new CollectionRenderer(
          component,
          this.componentMetadata,
          this.importCollection,
          this.renderConfig,
          parent,
        ).renderElement(renderChildren);

      case Primitive.Divider:
        return new ReactComponentRenderer<DividerProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement();

      case Primitive.Expander:
        return new ReactComponentRenderer<ExpanderProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.ExpanderItem:
        return new ReactComponentRenderer<ExpanderItemProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.Flex:
        return new ReactComponentRenderer<FlexProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.Grid:
        return new ReactComponentRenderer<GridProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.Heading:
        return new ReactComponentRenderer<HeadingProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.Icon:
        return new ReactComponentRenderer<IconProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.Image:
        return new ReactComponentRenderer<ImageProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement();

      case Primitive.Link:
        return new ReactComponentRenderer<LinkProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.Loader:
        return new ReactComponentRenderer<LoaderProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.MenuButton:
        return new ReactComponentRenderer<ButtonProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.MenuItem:
        return new ReactComponentRenderer<MenuItemProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.Menu:
        return new ReactComponentRenderer<MenuProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.Pagination:
        return new ReactComponentRenderer<PaginationProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.PasswordField:
        return new ReactComponentRenderer<PasswordFieldProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.PhoneNumberField:
        return new ReactComponentRenderer<PhoneNumberFieldProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.Placeholder:
        return new ReactComponentRenderer<PlaceholderProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.Radio:
        return new ReactComponentRenderer<RadioProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.RadioGroupField:
        return new ReactComponentRenderer<RadioGroupFieldProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.Rating:
        return new ReactComponentRenderer<RatingProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.ScrollView:
        return new ReactComponentRenderer<ScrollViewProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.SearchField:
        return new ReactComponentRenderer<SearchFieldProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.SelectField:
        return new ReactComponentRenderer<SelectFieldProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.SliderField:
        return new ReactComponentRenderer<SliderFieldProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement();

      case Primitive.StepperField:
        return new ReactComponentRenderer<StepperFieldProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.SwitchField:
        return new ReactComponentRenderer<SwitchFieldProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.TabItem:
        return new ReactComponentRenderer<TabItemProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.Tabs:
        return new ReactComponentRenderer<TabsProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.Table:
        return new ReactComponentRenderer<TableProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.TableBody:
        return new ReactComponentRenderer<TableBodyProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.TableCell:
        return new ReactComponentRenderer<TableCellProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.TableFoot:
        return new ReactComponentRenderer<TableFootProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.TableHead:
        return new ReactComponentRenderer<TableHeadProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.TableRow:
        return new ReactComponentRenderer<TableRowProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.Text:
        return new ReactComponentRenderer<TextProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.TextAreaField:
        return new ReactComponentRenderer<TextAreaFieldProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.TextField:
        return new ReactComponentRenderer<TextFieldProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.ToggleButton:
        return new ReactComponentRenderer<ToggleButtonProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.ToggleButtonGroup:
        return new ReactComponentRenderer<ToggleButtonGroupProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.View:
        return new ReactComponentRenderer<ViewProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      case Primitive.VisuallyHidden:
        return new ReactComponentRenderer<VisuallyHiddenProps>(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);

      default:
        return new CustomComponentRenderer(
          component,
          this.componentMetadata,
          this.importCollection,
          parent,
        ).renderElement(renderChildren);
    }
  }
}
