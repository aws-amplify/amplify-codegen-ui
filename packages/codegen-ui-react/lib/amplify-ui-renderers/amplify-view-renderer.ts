/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */

import { JsxElement, factory, JsxFragment } from 'typescript';
import { ReactViewTemplateRenderer } from '../views/react-view-renderer';
import { Primitive } from '../primitive';
import { ReactTableRenderer } from '../react-table-renderer';

export class AmplifyViewRenderer extends ReactViewTemplateRenderer {
  renderJsx(): JsxElement | JsxFragment {
    switch (this.viewComponent.viewConfiguration.type) {
      case Primitive.Table:
        return new ReactTableRenderer(
          this.viewComponent,
          this.viewDefinition,
          this.viewMetadata,
          this.importCollection,
        ).renderElement();
      default:
        return factory.createJsxFragment(factory.createJsxOpeningFragment(), [], factory.createJsxJsxClosingFragment());
    }
  }
}
