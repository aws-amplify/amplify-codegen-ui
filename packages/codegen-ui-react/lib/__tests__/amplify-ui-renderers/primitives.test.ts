/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import { StudioComponent } from '@aws-amplify/codegen-ui';
import { assertASTMatchesSnapshot } from '../__utils__';
import { AmplifyRenderer } from '../../amplify-ui-renderers/amplify-renderer';
import { Primitive } from '../../primitive';

function testPrimitive(component: StudioComponent) {
  const renderedComponent = new AmplifyRenderer(component, {}).renderJsx(component);
  assertASTMatchesSnapshot(renderedComponent);
}

describe('Primitives', () => {
  Object.values(Primitive).forEach((primitive) => {
    test(primitive, () => {
      testPrimitive({
        componentType: primitive,
        name: `My${primitive}`,
        properties: {},
        bindingProperties: {},
        schemaVersion: '1.0',
      });
    });
  });
});
