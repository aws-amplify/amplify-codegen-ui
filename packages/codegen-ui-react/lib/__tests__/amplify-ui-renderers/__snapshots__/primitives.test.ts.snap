// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Primitives Alert 1`] = `"<Alert {...getOverrideProps(overrides, \\"MyAlert\\")} {...rest}></Alert>"`;

exports[`Primitives Autocomplete 1`] = `"<Autocomplete {...getOverrideProps(overrides, \\"MyAutocomplete\\")} {...rest}></Autocomplete>"`;

exports[`Primitives Badge 1`] = `"<Badge {...getOverrideProps(overrides, \\"MyBadge\\")} {...rest}></Badge>"`;

exports[`Primitives Button 1`] = `"<Button {...getOverrideProps(overrides, \\"MyButton\\")} {...rest}></Button>"`;

exports[`Primitives ButtonGroup 1`] = `"<ButtonGroup {...getOverrideProps(overrides, \\"MyButtonGroup\\")} {...rest}></ButtonGroup>"`;

exports[`Primitives Card 1`] = `"<Card {...getOverrideProps(overrides, \\"MyCard\\")} {...rest}></Card>"`;

exports[`Primitives CheckboxField 1`] = `"<CheckboxField {...getOverrideProps(overrides, \\"MyCheckboxField\\")} {...rest}></CheckboxField>"`;

exports[`Primitives Collection 1`] = `"<Collection items={items || []} {...getOverrideProps(overrides, \\"MyCollection\\")} {...rest}>{(item, index) => ()}</Collection>"`;

exports[`Primitives Divider 1`] = `"<Divider {...getOverrideProps(overrides, \\"MyDivider\\")} {...rest}></Divider>"`;

exports[`Primitives Expander 1`] = `"<Expander {...getOverrideProps(overrides, \\"MyExpander\\")} {...rest}></Expander>"`;

exports[`Primitives ExpanderItem 1`] = `"<ExpanderItem {...getOverrideProps(overrides, \\"MyExpanderItem\\")} {...rest}></ExpanderItem>"`;

exports[`Primitives Flex 1`] = `"<Flex {...getOverrideProps(overrides, \\"MyFlex\\")} {...rest}></Flex>"`;

exports[`Primitives Grid 1`] = `"<Grid {...getOverrideProps(overrides, \\"MyGrid\\")} {...rest}></Grid>"`;

exports[`Primitives Heading 1`] = `"<Heading {...getOverrideProps(overrides, \\"MyHeading\\")} {...rest}></Heading>"`;

exports[`Primitives Icon 1`] = `"<Icon {...getOverrideProps(overrides, \\"MyIcon\\")} {...rest}></Icon>"`;

exports[`Primitives Image 1`] = `"<Image {...getOverrideProps(overrides, \\"MyImage\\")} {...rest}></Image>"`;

exports[`Primitives Link 1`] = `"<Link {...getOverrideProps(overrides, \\"MyLink\\")} {...rest}></Link>"`;

exports[`Primitives Loader 1`] = `"<Loader {...getOverrideProps(overrides, \\"MyLoader\\")} {...rest}></Loader>"`;

exports[`Primitives Menu 1`] = `"<Menu {...getOverrideProps(overrides, \\"MyMenu\\")} {...rest}></Menu>"`;

exports[`Primitives MenuButton 1`] = `"<MenuButton {...getOverrideProps(overrides, \\"MyMenuButton\\")} {...rest}></MenuButton>"`;

exports[`Primitives MenuItem 1`] = `"<MenuItem {...getOverrideProps(overrides, \\"MyMenuItem\\")} {...rest}></MenuItem>"`;

exports[`Primitives Pagination 1`] = `"<Pagination {...getOverrideProps(overrides, \\"MyPagination\\")} {...rest}></Pagination>"`;

exports[`Primitives PasswordField 1`] = `"<PasswordField {...getOverrideProps(overrides, \\"MyPasswordField\\")} {...rest}></PasswordField>"`;

exports[`Primitives PhoneNumberField 1`] = `"<PhoneNumberField {...getOverrideProps(overrides, \\"MyPhoneNumberField\\")} {...rest}></PhoneNumberField>"`;

exports[`Primitives Placeholder 1`] = `"<Placeholder {...getOverrideProps(overrides, \\"MyPlaceholder\\")} {...rest}></Placeholder>"`;

exports[`Primitives Radio 1`] = `"<Radio {...getOverrideProps(overrides, \\"MyRadio\\")} {...rest}></Radio>"`;

exports[`Primitives RadioGroupField 1`] = `"<RadioGroupField {...getOverrideProps(overrides, \\"MyRadioGroupField\\")} {...rest}></RadioGroupField>"`;

exports[`Primitives Rating 1`] = `"<Rating {...getOverrideProps(overrides, \\"MyRating\\")} {...rest}></Rating>"`;

exports[`Primitives ScrollView 1`] = `"<ScrollView {...getOverrideProps(overrides, \\"MyScrollView\\")} {...rest}></ScrollView>"`;

exports[`Primitives SearchField 1`] = `"<SearchField {...getOverrideProps(overrides, \\"MySearchField\\")} {...rest}></SearchField>"`;

exports[`Primitives SelectField 1`] = `"<SelectField {...getOverrideProps(overrides, \\"MySelectField\\")} {...rest}></SelectField>"`;

exports[`Primitives SliderField 1`] = `"<SliderField {...getOverrideProps(overrides, \\"MySliderField\\")} {...rest}></SliderField>"`;

exports[`Primitives StepperField 1`] = `"<StepperField {...getOverrideProps(overrides, \\"MyStepperField\\")} {...rest}></StepperField>"`;

exports[`Primitives SwitchField 1`] = `"<SwitchField {...getOverrideProps(overrides, \\"MySwitchField\\")} {...rest}></SwitchField>"`;

exports[`Primitives TabItem 1`] = `"<TabItem {...getOverrideProps(overrides, \\"MyTabItem\\")} {...rest}></TabItem>"`;

exports[`Primitives Table 1`] = `"<Table {...getOverrideProps(overrides, \\"MyTable\\")} {...rest}></Table>"`;

exports[`Primitives TableBody 1`] = `"<TableBody {...getOverrideProps(overrides, \\"MyTableBody\\")} {...rest}></TableBody>"`;

exports[`Primitives TableCell 1`] = `"<TableCell {...getOverrideProps(overrides, \\"MyTableCell\\")} {...rest}></TableCell>"`;

exports[`Primitives TableFoot 1`] = `"<TableFoot {...getOverrideProps(overrides, \\"MyTableFoot\\")} {...rest}></TableFoot>"`;

exports[`Primitives TableHead 1`] = `"<TableHead {...getOverrideProps(overrides, \\"MyTableHead\\")} {...rest}></TableHead>"`;

exports[`Primitives TableRow 1`] = `"<TableRow {...getOverrideProps(overrides, \\"MyTableRow\\")} {...rest}></TableRow>"`;

exports[`Primitives Tabs 1`] = `"<Tabs {...getOverrideProps(overrides, \\"MyTabs\\")} {...rest}></Tabs>"`;

exports[`Primitives Text 1`] = `"<Text {...getOverrideProps(overrides, \\"MyText\\")} {...rest}></Text>"`;

exports[`Primitives TextAreaField 1`] = `"<TextAreaField {...getOverrideProps(overrides, \\"MyTextAreaField\\")} {...rest}></TextAreaField>"`;

exports[`Primitives TextField 1`] = `"<TextField {...getOverrideProps(overrides, \\"MyTextField\\")} {...rest}></TextField>"`;

exports[`Primitives ToggleButton 1`] = `"<ToggleButton {...getOverrideProps(overrides, \\"MyToggleButton\\")} {...rest}></ToggleButton>"`;

exports[`Primitives ToggleButtonGroup 1`] = `"<ToggleButtonGroup {...getOverrideProps(overrides, \\"MyToggleButtonGroup\\")} {...rest}></ToggleButtonGroup>"`;

exports[`Primitives View 1`] = `"<View {...getOverrideProps(overrides, \\"MyView\\")} {...rest}></View>"`;

exports[`Primitives VisuallyHidden 1`] = `"<VisuallyHidden {...getOverrideProps(overrides, \\"MyVisuallyHidden\\")} {...rest}></VisuallyHidden>"`;
